File: C++/Linked List/Doubly Circular Linked List/DoublyCL04.cpp
/*
===============================================================================
Program Name    : Doubly Circular Linked List - Display
Program Number  : 004
File Name       : DoublyCL04.cpp

Description     : This program demonstrates how to display all
                  elements of a Doubly Circular Linked List using
                  the Display() function.

Date            : August 2026
Language        : C++
Author          : Atharv Tushar Bhosale
===============================================================================

Concepts Used:
1. Class
2. Dynamic Memory Allocation
3. Self-Referential Class
4. Doubly Circular Linked List
5. InsertFirst Operation
6. InsertLast Operation
7. Display Operation

===============================================================================
*/

#include<iostream>
using namespace std;

class Node
{
    public:
        int data;
        Node *next;
        Node *prev;
};

class DoublyCL
{
    private:
        Node *First;
        Node *Last;
        int iCount;

    public:
        DoublyCL()
        {
            First = NULL;
            Last = NULL;
            iCount = 0;
        }

        void InsertFirst(int no)
        {
            Node *newn = new Node;

            newn->data = no;
            newn->next = NULL;
            newn->prev = NULL;

            if((First == NULL) && (Last == NULL))
            {
                First = Last = newn;
            }
            else
            {
                newn->next = First;
                First->prev = newn;
                First = newn;
            }

            Last->next = First;
            First->prev = Last;

            iCount++;
        }

        void InsertLast(int no)
        {
            Node *newn = new Node;

            newn->data = no;
            newn->next = NULL;
            newn->prev = NULL;

            if((First == NULL) && (Last == NULL))
            {
                First = Last = newn;
            }
            else
            {
                Last->next = newn;
                newn->prev = Last;
                Last = newn;
            }

            Last->next = First;
            First->prev = Last;

            iCount++;
        }

        void Display()
        {
            if((First == NULL) && (Last == NULL))
            {
                cout<<"Linked List is Empty\n";
                return;
            }

            Node *temp = First;

            cout<<"<=> ";

            do
            {
                cout<<"| "<<temp->data<<" | <=> ";
                temp = temp->next;
            }while(temp != First);

            cout<<"\n";
        }
};

int main()
{
    DoublyCL dobj;

    dobj.InsertFirst(21);
    dobj.InsertFirst(11);

    dobj.InsertLast(51);
    dobj.InsertLast(101);

    cout<<"Elements of Doubly Circular Linked List are:\n";
    dobj.Display();

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Elements of Doubly Circular Linked List are:

<=> | 11 | <=> | 21 | <=> | 51 | <=> | 101 | <=>

===============================================================================
Memory Representation:
===============================================================================

+----+ <=> +----+ <=> +----+ <=> +-----+
| 11 |     | 21 |     | 51 |     | 101 |
+----+ <=> +----+ <=> +----+ <=> +-----+

First = 11
Last  = 101

101.next = 11
11.prev = 101

===============================================================================
Algorithm:
===============================================================================

Display()

1. Check whether the list is empty.
2. Create a temporary pointer and initialize it to First.
3. Display node data.
4. Move temp to next node.
5. Repeat until temp reaches First again.
6. Stop traversal.

===============================================================================
Time Complexity:
===============================================================================

InsertFirst() : O(1)

InsertLast()  : O(1)

Display()     : O(n)

Space Complexity : O(1)

where,
n = Number of Nodes

===============================================================================
Conclusion:
===============================================================================

Program 004 demonstrates traversal and display of all nodes
present in a Doubly Circular Linked List. Since the list is
circular, traversal continues until the First node is reached
again.

===============================================================================
*/

File: Java/Linked List/Doubly Circular Linked List/DoublyCL04.java
/*
===============================================================================
Program Name    : Doubly Circular Linked List - Display Function
Program Number  : 004
File Name       : DoublyCL04.java

Description     : This program demonstrates the Display()
                  operation in a Doubly Circular Linked List
                  using Java. The linked list is traversed
                  from the first node to the last node and
                  all elements are displayed. 

Date            : July 2026
Language        : Java
Author          : Atharv Tushar Bhosale
===============================================================================

Concepts Used:
1. Class and Object
2. Self-Referential Class
3. Doubly Circular Linked List
4. Dynamic Memory Allocation
5. do-while Loop
6. Circular Traversal

Key Learning Points:
- Traverse a Doubly Circular Linked List.
- Display all nodes sequentially.
- Use do-while loop for circular traversal.
- Maintain circular links between first and last nodes.

===============================================================================
*/

class node
{
    public int data;
    public node next;
    public node prev;

    node(int no)
    {
        this.data = no;
        this.next = null;
        this.prev = null;
    }
}

class DoublyCL
{
    private node first;
    private node last;
    private int iCount;

    public DoublyCL()
    {
        first = null;
        last = null;
        iCount = 0;
    }

    public void Display()
    {
        if(first == null && last == null)
        {
            return;
        }

        node temp = first;

        do
        {
            System.out.print("| " + temp.data + " | <=> ");
            temp = temp.next;
        }while(temp != last.next);

        System.out.println();
    }

    public void InsertFirst(int iNo)
    {
        node newn = new node(iNo);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            newn.next = first;
            first.prev = newn;
            first = newn;
        }

        last.next = first;
        first.prev = last;

        iCount++;
    }

    public void InsertLast(int iNo)
    {
        node newn = new node(iNo);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            last.next = newn;
            newn.prev = last;
            last = newn;
        }

        last.next = first;
        first.prev = last;

        iCount++;
    }
}

class DoublyCL04
{
    public static void main(String A[])
    {
        DoublyCL dobj = new DoublyCL();

        dobj.InsertFirst(51);
        dobj.InsertFirst(21);
        dobj.InsertFirst(11);

        dobj.InsertLast(101);
        dobj.InsertLast(111);

        dobj.Display();
    }
}

/*
===============================================================================
Expected Output:
===============================================================================

| 11 | <=> | 21 | <=> | 51 | <=> | 101 | <=> | 111 | <=>

===============================================================================
Memory Representation:
===============================================================================

 first
   |
   v

+----+ <=> +----+ <=> +----+ <=> +-----+ <=> +-----+
| 11 |     | 21 |     | 51 |     | 101 |     | 111 |
+----+ <=> +----+ <=> +----+ <=> +-----+ <=> +-----+
   ^                                                 |
   |_________________________________________________|

===============================================================================
Algorithm:
===============================================================================

1. Start from the first node.
2. Display node data.
3. Move to the next node.
4. Repeat until traversal reaches first node again.
5. Stop.

===============================================================================
Time Complexity:
===============================================================================

Display()     : O(n)

InsertFirst() : O(1)

InsertLast()  : O(1)

Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 004 demonstrates the Display() operation of a Doubly
Circular Linked List. The do-while loop ensures that every
node is displayed exactly once while preserving the circular
nature of the linked list.

===============================================================================
*/
